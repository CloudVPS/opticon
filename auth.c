#include <auth.h>

// http://www.literatecode.com/get/aes256.c

/** Initialize session-related global storage. */
void sessionlist_init (void) {
	for (int i=0; i<256; ++i) {
		SESSIONS[i].first = SESSIONS[i].last = NULL;
	}
}

/** Allocate and initialize session object. */
session *session_alloc (void) {
	session *res = (session *) malloc (sizeof (session));
	res->next = res->prev = NULL;
	res->addr = res->sessid = 0;
	res->key.data = {0,0,0,0,0,0,0,0};
	res->lastcycle = time (NULL);
	res->host = NULL;
	return res;
}

/** Link a new session into the proper bucket. */
void session_link (session *s) {
	uint32_t sid = s->sessid ^ s->addr;
	uint8_t bucket = (sid >> 24) |
					 ((sid & 0x00ff0000) >> 16) |
					 ((sid & 0x0000ff00) >> 8) |
					 (sid & 0xff);
	if (SESSIONS[bucket].first) {
		// link in at top.
		s->next = SESSIONS[bucket].first;
		SESSIONS[bucket].first->prev = s;
		SESSIONS[bucket].first = s;
	}
	else {
		SESSIONS[bucket].first = SESSIONS[bucket].last = s;
	}
}

/** Register a new session (or update an existing session to
  * stop it from expiring. */
int session_register (uuid tenantid, uuid hostid,
					  uint32_t addrpart, uint32_t sess_id.
					  aeskey sess_key) {
	host *h = host_find (tenantid, hostid);
	session *s;
	if (! h) return 0;

	if (s = session_find (hostid, addrpart)) {
		s->lastcycle = time (NULL);
		return 1;
	}
		
	s = session_alloc();
	if (! s) return 0;
	
	s->tenantid = tenantid;
	s->hostid = hostid;
	s->addr = addrpart;
	s->sessid = sess_id;
	s->key = sess_key;
	
	session_link (s);
	return 1;
}

session *session_find (uint32_t addrpart, uint32_t sess_id) {
	uint32_t sid = s->sessid ^ s->addr;
	uint8_t bucket = (sid >> 24) |
					 ((sid & 0x00ff0000) >> 16) |
					 ((sid & 0x0000ff00) >> 8) |
					 (sid & 0xff);
	session *s = SESSIONS[bucket].first;
	while (s) {
		if (s->addr == addrpart && s->sessid == sess_id) return s;
		s = s->next;
	}
	return NULL;
}

void session_expire (time_t cutoff) {
	session *s;
	for (uint8_t i=0; i<=255; ++i) {
		s = SESSIONS[i].first;
		while (s) {
			if (s->lastcycle < cutoff) {
				if (s->prev) {
					if (s->next) {
						s->prev->next = s->next;
						s->next->prev = s->prev;
					}
					else {
						s->prev->next = NULL;
						SESSIONS[i].last = s->prev;
					}
				}
				else {
					if (s->next) {
						s->next->prev = NULL;
						SESSIONS[i].first = s->next;
					}
					else {
						SESSIONS[i].first = SESSIONS[i].last = NULL;
					}
				}
				
				free (s);
			}
			s = s->next;
		}
	}
}
