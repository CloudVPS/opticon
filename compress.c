#include <compress.h>

int compress_data (ioport *in, ioport *out) {
    if (ioport_read_available (in) == 0) return 0;
    
    z_stream strm;
    strm.zalloc = Z_NULL;
    strm.zfree = Z_NULL;
    strm.opaque = Z_NULL;
    strm.total_out = 0;
    strm.next_in = (Bytef*)ioport_get_buffer (in);
    strm.avail_in = ioport_read_available (in);
    
    int r = deflateInit2 (&strm, Z_DEFAULT_COMPRESSION,
                          Z_DEFLATED, 15+16, 8,
                          Z_DEFAULT_STRATEGY);
    if (r != Z_OK) return 0;
    
    int st;
    do {
        strm.next_out = (Bytef*)ioport_get_buffer (out) + strm.total_out;
        strm.avail_out = ioport_write_available (out) - strm.total_out;
        st = deflate (&strm, Z_FINISH);
    } while (st == Z_OK);
    
    if (st != Z_STREAM_END) return 0;
    ioport_write (out, ioport_get_buffer (out), strm.total_out);
    return 1;
}

int decompress_data (ioport *in, ioport *out) {
    return 0;
}
