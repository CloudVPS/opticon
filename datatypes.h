#ifndef _DATATYPES_H
#define _DATATYPES_H 1

#include <stdint.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>

typedef uint64_t		meterid_t;
typedef uint64_t		metertype_t;
typedef uint32_t		status_t;

#define MTYPE_INT 		0x0000000000000000
#define MTYPE_FRAC		0x4000000000000000
#define MTYPE_STR   	0x8000000000000000

#define MMASK_TYPE		0xc000000000000000
#define MMASK_COUNT		0x0000000000000007
#define MMASK_NAME		0x3ffffffffffffff8

meterid_t 	makeid (const char *label, metertype_t type, int pos);
void	  	id2label (meterid_t id, char *into);
metertype_t	id2type (meterid_t id);

typedef struct { uint64_t msb; uint64_t lsb; } uuid_t;

typedef union {
	uint64_t		 *u64;
	double			 *frac;
	char			 *str;
	void			 *any;
} meterdata;

typedef struct meter_s {
	struct meter_s	*next;
	struct meter_s	*prev;

	meterid_t		 id;
	time_t			 lastmodified;
	int				 count;
	meterdata		 d;
} meter;

typedef struct host_s {
	struct host_s	*next;
	struct host_s	*prev;
	uuid_t			 uuid;
	status_t		 status;
	meter			*first;
	meter			*last;
} host;	

typedef struct tenant_s {
	struct tenant_s	*next;
	struct tenant_s	*prev;
	host			*first;
	host			*last;
	uuid_t			 uuid;
	char			*key;
} tenant;

typedef struct {
	tenant			*first;
	tenant			*last;
} tenantlist;

extern tenantlist TENANTS;

tenant *tenant_find (uuid_t tenantid);
tenant *tenant_create (uuid_t tenantid, const char *key);

host *host_alloc (void);
host *host_find (uuid_t tenantid, uuid_t hostid);

meter *host_get_meter (host *h, meterid_t id);
meter *host_set_meter_uint (host *h, meterid_t id, 
							unsigned int count,
							uint64_t *data);
meter *host_set_meter_frac (host *h, meterid_t id,
							unsigned int count,
							double *data);
meter *host_set_meter_str  (host *h, meterid_t id, const char *str);

uint32_t	meter_get_uint (meter *, unsigned int pos);
double		meter_get_frac (meter *, unsigned int pos);
const char *meter_get_str (meter *);

#endif
