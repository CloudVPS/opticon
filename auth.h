#ifndef _AUTH_H
#define _AUTH_H 1

#include <datatypes.h>
#include <time.h>

typedef struct aeskey_s {
    uint8_t data[32];
} aeskey;

typedef struct session_s {
    struct session_s    *next;
    struct session_s    *prev;
    uuid                 tenantid;
    uuid                 hostid;
    uint32_t             addr;
    uint32_t             sessid;
    aeskey               key;
    time_t               lastcycle;
    host                *host;
} session;

typedef struct sessionlist_s {
    session             *first;
    session             *last;
} sessionlist;

extern sessionlist SESSIONS[256];

aeskey       aeskey_create (void);

void         sessionlist_init (void);
session     *session_alloc (void);
void         session_link (session *);

session     *session_register (uuid tenantid, uuid hostid, 
                               uint32_t addrpart, uint32_t sess_id,
                               aeskey sess_key);
                               
session     *session_find (uint32_t addr, uint32_t sess_id);
void         session_expire (time_t);

#endif
