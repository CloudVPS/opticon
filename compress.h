#ifndef _COMPRESS_H
#define _COMPRESS_H 1

#include <datatypes.h>
#include <ioport.h>
#include <zlib.h>

/* ============================= FUNCTIONS ============================= */

int compress_data (ioport *in, ioport *out);
int decompress_data (ioport *in, ioport *out);

#endif
