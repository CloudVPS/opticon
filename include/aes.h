#ifndef _AES_H
#define _AES_H 1

#include <stdlib.h>
#include <datatypes.h>
#include <ioport.h>
#include <auth.h>

/* =============================== TYPES =============================== */

typedef struct {
    uint8_t key[32]; 
    uint8_t enckey[32]; 
    uint8_t deckey[32];
} aes256_context; 

/* ============================= FUNCTIONS ============================= */

void aes256_init(aes256_context *, uint8_t * /* key */);
void aes256_done(aes256_context *);
void aes256_encrypt_ecb(aes256_context *, uint8_t * /* plaintext */);
void aes256_decrypt_ecb(aes256_context *, uint8_t * /* cipertext */);

int ioport_encrypt (aeskey *k, ioport *in, ioport *out, time_t ts);
int ioport_decrypt (aeskey *k, ioport *in, ioport *out, time_t ts);

#endif
