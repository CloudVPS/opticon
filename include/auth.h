#ifndef _AUTH_H
#define _AUTH_H 1

#include <datatypes.h>
#include <ioport.h>
#include <time.h>

/* =============================== TYPES =============================== */

/** Storage for an AES256 key */
typedef struct aeskey_s {
    uint8_t data[32];
} aeskey;

/** Representation of an active session */
typedef struct session_s {
    struct session_s    *next; /**< List link */
    struct session_s    *prev; /**< List link */
    uuid                 tenantid; /**< Tenant uuid associated */
    uuid                 hostid; /**< Host uuid associated */
    uint32_t             addr; /**< IP-derived id */
    uint32_t             sessid; /**< The sessionid */
    aeskey               key; /**< Session AES key */
    time_t               lastcycle; /**< Time since last key cycle */
    host                *host; /**< Connected host */
} session;

/** Linked list head for sessions */
typedef struct sessionlist_s {
    session             *first; /**< This bucket's first session */
    session             *last; /**< This bucket's last session */
} sessionlist;

/* ============================== GLOBALS ============================== */

extern sessionlist SESSIONS[256];

/* ============================= FUNCTIONS ============================= */

aeskey       aeskey_create (void);

void         sessionlist_init (void);
session     *session_alloc (void);
void         session_link (session *);

session     *session_register (uuid tenantid, uuid hostid, 
                               uint32_t addrpart, uint32_t sess_id,
                               aeskey sess_key);
                               
session     *session_find (uint32_t addr, uint32_t sess_id);
void         session_expire (time_t);
void         session_print (session *, ioport *);

#endif
