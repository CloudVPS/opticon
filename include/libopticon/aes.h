#ifndef _AES_H
#define _AES_H 1

#include <stdlib.h>
#include <sys/types.h>
#include <libopticon/datatypes.h>
#include <libopticon/ioport.h>

/* =============================== TYPES =============================== */

typedef struct {
    uint8_t key[32]; 
    uint8_t enckey[32]; 
    uint8_t deckey[32];
} aes256_context; 

/* ============================= FUNCTIONS ============================= */

void     aes256_init(aes256_context *, uint8_t * /* key */);
void     aes256_done(aes256_context *);
void     aes256_encrypt_ecb(aes256_context *, uint8_t * /* plaintext */);
void     aes256_decrypt_ecb(aes256_context *, uint8_t * /* ciphertext */);

aeskey   aeskey_create (void);
aeskey   aeskey_from_base64 (const char *);
char    *aeskey_to_base64 (aeskey);
int      ioport_encrypt (aeskey *k, ioport *in, 
                         ioport *out, time_t, uint32_t);
int      ioport_decrypt (aeskey *k, ioport *in, 
                         ioport *out, time_t, uint32_t);

#endif
