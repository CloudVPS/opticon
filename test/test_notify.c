#include <libopticon/datatypes.h>
#include <libopticon/util.h>
#include <libopticon/auth.h>
#include <libopticon/codec_pkt.h>
#include <libopticon/codec_json.h>
#include <libopticon/compress.h>
#include <libopticon/ioport_file.h>
#include <libopticon/aes.h>
#include <libopticon/log.h>
#include <assert.h>
#include <stdio.h>

#define FD_STDOUT 1

void hexdump_buffer (ioport *io) {
    size_t sz = ioport_read_available (io);
    size_t i;
    char *buf = ioport_get_buffer (io);
    if (! buf) return;
    
    for (i=0; i<sz; ++i) {
        printf ("%02x ", (uint8_t) buf[i]);
        if ((i&31) == 31) printf ("\n");
    }
    if (i&31) printf ("\n");
}

int main (int argc, const char *argv[]) {
    const char *stenantid = "001b71534f4b4f1cb281cc06b134f98f";
    const char *shostid = "6f943a0d-bcd9-42fa-b0c9-6ede92f9a46a";
    char tstr[16];
    time_t tnow = time (NULL);
    int i;
    uuid tenantid = mkuuid (stenantid);
    uuid hostid = mkuuid (shostid);
    tenant_init();
    sessionlist_init();
    
    aeskey tenantkey = aeskey_create();
    tenant *T = tenant_create (tenantid, tenantkey);
    tenant_done (T);
    
    T = tenant_find (tenantid, TENANT_LOCK_WRITE);
    assert (T->uuid.msb == 0x001b71534f4b4f1c);
    assert (T->uuid.lsb == 0xb281cc06b134f98f);

    tenant_set_notification (T, true, "test", "ALERT", hostid);
    assert (notifylist_check_actionable (&T->notify) == false);
    
    /* hack timing */
    assert (T->notify.first);
    T->notify.first->lastchange = time (NULL) - 300;
    
    assert (notifylist_check_actionable (&T->notify));
    
    tenant_delete (T);
    return 0;
}
