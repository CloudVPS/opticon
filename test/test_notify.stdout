{
    "issues": {
        "6f943a0d-bcd9-42fa-b0c9-6ede92f9a46a": {
            "status": "ALERT",
            "isproblem": 1
        }
    }
}
